### PR TITLE
fix(files): make sure method is callable on a concrete object instance

### LINKED ISSUE #9010: Issues with ElggFile::detectMimeType()
I am seeing strange behaviour ever since we introduced https://github.com/Elgg/Elgg/pull/8889.

When calling `ElggFile::detectMimeType()` from a contcrete class instance (other than ElggFile), `$this` is considered set and refers to the concrete instance. As a result the script is WSODing, as `getFilenameOnFilestore` does not exist for that concrete class. 

We should deprecate the static use, and add another static method. In the meantime, we should perhaps make sure that `$this` is an instance of `ElggFile`

### LINKED ISSUE #9019: Integer casting in relationships API
Had a typo in my code and was surprised to see a relationship added with the site entity:

``` php
add_entity_relationship('something true', 'relationship', $guid_two);
```

### PR DESCRIPTION
**Pull request by hypeJunction**
PHP treats $this as an instance of a calling class when ElggFile::detectMimeType is called
statically.

Fixes #9010

### CODE REVIEW

File: engine/classes/ElggFile.php [hunk 1]
```diff
@@ -133,7 +133,7 @@ public function setMimeType($mimetype) {
 	 * @todo Move this out into a utility class
 	 */
 	public function detectMimeType($file = null, $default = null) {
-		if (!$file && isset($this)) {
+		if (!$file && isset($this) && is_callable(array($this, 'getFilenameOnFilestore'))) {
```
Comment (mrclay) [inline engine/classes/ElggFile.php]: Why not `$this instanceof __CLASS__`?

**Comment by hypeJunction** [comment]
@mrclay Updated as per your suggestion, though the construct `$this instanceof __CLASS__` is illegal syntax (apparently).

File: engine/classes/ElggFile.php [hunk 2]
```diff
@@ -133,7 +133,8 @@ public function setMimeType($mimetype) {
 	 * @todo Move this out into a utility class
 	 */
 	public function detectMimeType($file = null, $default = null) {
-		if (!$file && isset($this)) {
+		$class = __CLASS__;
```
Comment (mrclay) [inline engine/classes/ElggFile.php]: LOL PHP

File: engine/tests/ElggCoreFilestoreTest.php [hunk 3]
```diff
@@ -97,6 +97,34 @@ function testElggGetFileSimpletype() {
...
+
+		// mime should not be null if default is set
+		$this->assertTrue(isset($mime));
+
+		// mime of a file object should match mime of a file path that represents this file on filestore
+		$resource_mime = $file->detectMimeType($file->getFilenameOnFilestore(), 'text/plain');
+		$this->assertIdentical($mime, $resource_mime);
+
+		// method output should not differ between a static and a concrete call if the file path is set
+		$resource_mime_static = ElggFile::detectMimeType($file->getFilenameOnFilestore(), 'text/plain');
```
Comment (mrclay) [inline engine/tests/ElggCoreFilestoreTest.php]: PHPUnit raises warnings to exceptions, so I think you'll need something like:

before: `error_reporting(E_ALL & ~E_STRICT);`
after: `error_reporting(E_ALL);`

Comment (hypeJunction) [inline engine/tests/ElggCoreFilestoreTest.php]: Unless we declare detectMimeType as static again. It would make sense because it is a static function...

Comment (mrclay) [inline engine/tests/ElggCoreFilestoreTest.php]: In 1.12 I'd prefer not. Adding a 2nd method in 2.x would be OK.

**Comment by hypeJunction** [comment]
@Elgg/owners can we get some eyes on this and merge this in and up, it's causing headache

File: engine/classes/ElggFile.php (hunk 2 above)
Comment (hypeJunction) [inline engine/classes/ElggFile.php]: :crying_cat_face:

File: engine/tests/ElggCoreFilestoreTest.php (hunk 3 above)
Comment (hypeJunction) [inline engine/tests/ElggCoreFilestoreTest.php]: Still failing on 5.6

File: engine/tests/ElggCoreFilestoreTest.php [hunk 4]
```diff
@@ -97,6 +97,40 @@ function testElggGetFileSimpletype() {
...
+		// mime should not be null if default is set
+		$this->assertTrue(isset($mime));
+
+		// mime of a file object should match mime of a file path that represents this file on filestore
+		$resource_mime = $file->detectMimeType($file->getFilenameOnFilestore(), 'text/plain');
+		$this->assertIdentical($mime, $resource_mime);
+
+		// calling detectMimeType statically raises strict policy warning
+		// @todo: remove this once a new static method has been implemented
+		error_reporting(E_ALL & ~E_STRICT);
```
Comment (mrclay) [inline engine/tests/ElggCoreFilestoreTest.php]: ... `& ~E_DEPRECATED`

**Comment by hypeJunction** [comment]
All seems green here

**Comment by hypeJunction** [comment]
I don't see these changes in neither 2.0 or 2.x. Were there merge conflicts and these were skipped or something?

**Comment by mrclay** [comment]
They made it.

**Pull request merged**